In [ ]:
cd drive

/content/drive


In [ ]:
cd MyDrive/

/content/drive/MyDrive


In [ ]:
cd Global\ Terrorism

/content/drive/MyDrive/Global Terrorism


In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import math
import warnings
warnings.filterwarnings("ignore")

In [ ]:
import pandas as pd
df= pd.read_csv("GLT.csv",encoding='ISO-8859-1',low_memory=False)

In [ ]:
df.head(5)

eventid  iyear  imonth  iday  ... INT_IDEO  INT_MISC INT_ANY  related
0  197000000001   1970       7     2  ...        0         0       0      NaN
1  197000000002   1970       0     0  ...        1         1       1      NaN
2  197001000001   1970       1     0  ...       -9         1       1      NaN
3  197001000002   1970       1     0  ...       -9         1       1      NaN
4  197001000003   1970       1     0  ...       -9         1       1      NaN

[5 rows x 135 columns]

# Selection of Columns to be used in the analysis

Columns with too many Missing Values will be very difficult to analyze. In this case, we will drop the column with Missing Value more than 50%.

In [ ]:
#Specifies the percent of data that is blank for each column
len_df = len(df)
percent_null = []
for item in df.columns:
    count_null = df[item].isnull().sum()
    percent_null.append(count_null/len_df*100)

In [ ]:
percent_column_null = pd.DataFrame({'column_name':df.columns, 'percent_null(%)':percent_null})
percent_column_null

column_name  percent_null(%)
0       eventid         0.000000
1         iyear         0.000000
2        imonth         0.000000
3          iday         0.000000
4    approxdate        94.914993
..          ...              ...
130     INT_LOG         0.000000
131    INT_IDEO         0.000000
132    INT_MISC         0.000000
133     INT_ANY         0.000000
134     related        86.219461

[135 rows x 2 columns]

Many columns turned out to have a Missing Value of more than 50%, some even more than 90%. We will not use these columns.

In [ ]:
#Drop Missing Value column> 50%
column_to_drop = []
for i in range(len(percent_column_null)):
    if percent_column_null['percent_null(%)'][i] > 50:
        column_to_drop.append(percent_column_null['column_name'][i])

In [ ]:
df2 = df.drop(column_to_drop, axis=1)

In [ ]:
len(df2.columns)

58

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 58 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   eventid           181691 non-null  int64  
 1   iyear             181691 non-null  int64  
 2   imonth            181691 non-null  int64  
 3   iday              181691 non-null  int64  
 4   extended          181691 non-null  int64  
 5   country           181691 non-null  int64  
 6   country_txt       181691 non-null  object 
 7   region            181691 non-null  int64  
 8   region_txt        181691 non-null  object 
 9   provstate         181270 non-null  object 
 10  city              181257 non-null  object 
 11  latitude          177135 non-null  float64
 12  longitude         177134 non-null  float64
 13  specificity       181685 non-null  float64
 14  vicinity          181691 non-null  int64  
 15  summary           115562 non-null  object 
 16  crit1             18

In [ ]:
# Make a list of column names used for analysis
column_to_use = [ 'latitude', 'longitude', 'nperpcap', 'nkill', 'nkillter',
       'nwound', 'nwoundte', 'eventid', 'iyear', 'imonth', 'iday', 'extended',
       'crit1', 'crit2', 'crit3', 'multiple', 'country', 'country_txt',
       'region', 'region_txt', 'city', 'specificity', 'attacktype1',
       'attacktype1_txt', 'success', 'suicide', 'weaptype1', 'weaptype1_txt',
       'targtype1', 'targtype1_txt', 'natlty1', 'natlty1_txt', 'individual',
       'gname', 'claimed', 'property', 'doubtterr','provstate']

In [ ]:
df2 = df[column_to_use]

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 38 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   latitude         177135 non-null  float64
 1   longitude        177134 non-null  float64
 2   nperpcap         112202 non-null  float64
 3   nkill            171378 non-null  float64
 4   nkillter         114733 non-null  float64
 5   nwound           165380 non-null  float64
 6   nwoundte         112548 non-null  float64
 7   eventid          181691 non-null  int64  
 8   iyear            181691 non-null  int64  
 9   imonth           181691 non-null  int64  
 10  iday             181691 non-null  int64  
 11  extended         181691 non-null  int64  
 12  crit1            181691 non-null  int64  
 13  crit2            181691 non-null  int64  
 14  crit3            181691 non-null  int64  
 15  multiple         181690 non-null  float64
 16  country          181691 non-null  int6

In [ ]:
df2.nunique()

latitude            48322
longitude           48039
nperpcap               50
nkill                 205
nkillter               96
nwound                238
nwoundte               44
eventid            181691
iyear                  47
imonth                 13
iday                   32
extended                2
crit1                   2
crit2                   2
crit3                   2
multiple                2
country               205
country_txt           205
region                 12
region_txt             12
city                36674
specificity             5
attacktype1             9
attacktype1_txt         9
success                 2
suicide                 2
weaptype1              12
weaptype1_txt          12
targtype1              22
targtype1_txt          22
natlty1               215
natlty1_txt           215
individual              2
gname                3537
claimed                 3
property                3
doubtterr               3
provstate            2855
dtype: int64

### **Explanation of Column Names Used**
There are 38 columns used in this analysis. The 38 columns are as follows.

•	eventid- this is the Unique ID of the event. (Numeric Variable)

•	iyear-year of the incident (Numeric Variable)

•	imonth-month of the incident (Numeric Variable)

•	iday – Day of the incident (Numeric Variable)

•	extended- Extended Incident (Categorical Variable)

•	summary -Incident Summary (Text Variable)

•	crit1, crit2, crit3-Inclusion Criteria (Categorical Variables)

•	multiple - part of multiple incident (Categorical Variable)

•	related - related incidents (Text Variable)

•	country, country_txt- country (Categorical Variable)

•	region, region_txt- region (Categorical Variable)

•	city - city (Text Variable)

•	latitude – latitude (Numeric Variable)

•	longitude - longitude (Numeric Variable)

•	specificity -Geo-Coding Specificity (Categorical Variable)

•	attacktype1, attacktype1_txt – Attack Type (Categorical Variable)

•	success -Successful Attack (Categorical Variable)

•	suicide -Suicide Attack (Categorical Variable)

•	weaptype1, weaptype1_txt -Weapon Type (Categorical Variable)

•	targtype1, targtype1_txt -Victim Type (Categorical Variable)

•	corp1: Name of the Entity (Text Variable)

•	natlty1; natlty1_txt -Nationality of the Victim (Categorical Variable)

•	individual -Unaffiliated Individuals (Categorical Variable)

•	gname -Perpetrator Group Name (Text Variable)

•	nperps- Number of the Perpetrators (Numeric Variable)

•	claimed -Claim of Responsibility (Categorical Variable)

•	motive - Motive (Text Variable)

•	Nkill - Total Fatalities - (Numeric Variable)

•	nkillter - perpetrator Fatalities (Numeric Variable)

•	nwound -Total Injured (Numeric Variable)

•	nwoundte - Perpetrators Injured (Numeric Variable)

•	property - property damage (Categorical Variable)

•	propvalue- Value of the Property Damage (in usd) (Numeric Variable)

•	nhostkid- Total No. Of Hostages / Victim Kidnaps (- Numeric Variable)

•	ransomamt -Total Ransom Amount Demanded (Numeric Variable)

•	ransompaid -Total Ransom Amount Paid (Numeric Variable)

•	INT_IDEO – Attack of an International Ideology (Categorical Variable)

•	doubtterr - Doubt Terrorism Proper (Categorical Variable)


# **Empty Data Handling**

In [ ]:
df2.isnull().sum()


There are empty data from the Province and City columns. We can fill this data based on Latitude and Longitude.

In [ ]:
#Create df_test_city which is City, Lat, and Long data that is not null
df_test_city = df2[(df2['city'].notnull()) & (df2['latitude'].notnull()) & (df2['longitude'].notnull())]

#Create df_predict_city which is city data that is empty, but Lat and Long are not null
df_predict_city = df2[(df2['city'].isnull()) & (df2['latitude'].notnull()) & (df2['longitude'].notnull())]

In [ ]:
len(df_predict_city)

426

The data values are empty in City but Latitude and Longitude are not empty. There are 426 data from 434 City data that are empty. This means we can fill almost all City data that is empty.

In [ ]:
def fill_city(df):
    lat_diff = np.abs(df['latitude'] - df_test_city['latitude'])
    long_diff = np.abs(df['longitude'] - df_test_city['longitude'])
    diff = lat_diff+long_diff
    index_min = np.argmin(diff,axis=0,out=None)
    return df_test_city['city'].loc[index_min]

In [ ]:
idx_city_change = df_predict_city.index
df_test_city=df_test_city.reset_index(drop=True)

In [ ]:
# Fill in the empty city data using apply
df2.loc[idx_city_change, 'city'] = df_predict_city.apply(fill_city, axis=1)

In [ ]:
df2.isnull().sum()

latitude            4556
longitude           4557
nperpcap           69489
nkill              10313
nkillter           66958
nwound             16311
nwoundte           69143
eventid                0
iyear                  0
imonth                 0
iday                   0
extended               0
crit1                  0
crit2                  0
crit3                  0
multiple               1
country                0
country_txt            0
region                 0
region_txt             0
city                   8
specificity            6
attacktype1            0
attacktype1_txt        0
success                0
suicide                0
weaptype1              0
weaptype1_txt          0
targtype1              0
targtype1_txt          0
natlty1             1559
natlty1_txt         1559
individual             0
gname                  0
claimed            66120
property               0
doubtterr              1
provstate            421
dtype: int64


8 City data remaining that are still empty. We will drop the data, because the amount is not too much.

In [ ]:
df2 = df2.dropna(subset=['city'])

# **Fill in Missing Value for Province**
Filling missing Province values almost the same as City, which is looking for the nearest latitude and longitude. Based on the information, the number of missing province values are 421.

In [ ]:
df_test_province = df2[(df2['provstate'].notnull()) & (df2['latitude'].notnull()) & (df2['longitude'].notnull())]
df_predict_province = df2[(df2['provstate'].isnull()) & (df2['latitude'].notnull()) & (df2['longitude'].notnull())]

In [ ]:
len(df_predict_province)

409

409 data to be filled out of 421 NaN Province data

In [ ]:
def fill_province(df):
    lat_diff = np.abs(df['latitude'] - df_test_province['latitude'])
    long_diff = np.abs(df['longitude'] - df_test_province['longitude'])
    diff = lat_diff+long_diff
    index_min = np.argmin(diff)
    return df_test_province['provstate'].loc[index_min]

In [ ]:
idx_prov_change = df_predict_province.index
df_test_province=df_test_province.reset_index(drop=True)

In [ ]:
df2.loc[idx_prov_change, 'provstate'] = df_predict_province.apply(fill_province, axis=1)

Similar to city data, we will also drop Provincial data which is still empty.

In [ ]:
df2 = df2.dropna(subset=['provstate'])

# **Fill Latitude and Longitude**

In [ ]:
df2[df2.latitude.isnull()]['city'].value_counts()

Unknown        2239
unknown          46
Nationwide       24
Nationwide*      17
Puerto Rico      14
               ... 
Yoque-Am          1
Vergel            1
Hisar             1
La Casita         1
Kupeli            1
Name: city, Length: 1859, dtype: int64

In [ ]:
df2[df2.latitude.isnull()]['country'].value_counts()

61     472
45     471
145    398
160    380
159    288
      ... 
164      1
162      1
197      1
406      1
22       1
Name: country, Length: 141, dtype: int64

Lines with Lat and Long are empty, most have City values ​​in the form Unknown. Therefore, we cannot use City to fill Lat and Long. Instead, the Country column is all filled in. So that we can fill Lat and Long based on the value from Country.

In [ ]:
mean_lat_country = df2.groupby('country').mean()['latitude']
mean_long_country = df2.groupby('country').mean()['longitude']

In [ ]:
# Create a function to fill in the blank latitude and longitude with the mean of the latitude or longitude of the country
def func_lat(df):
    if pd.isnull(df.latitude):
        return mean_lat_country[df.country]
    else:
        return df.latitude
    
def func_long(df):
    if pd.isnull(df.longitude):
        return mean_long_country[df.country]
    else:
        return df.longitude

In [ ]:
df2['latitude'] = df2.apply(func_lat, axis=1)

In [ ]:
df2['longitude'] = df2.apply(func_long, axis=1)

In [ ]:
# Delete Lat and Long data that is still empty
df2.dropna(subset=['latitude', 'longitude'], inplace=True)

The data that is still empty is now in the NatTarget, NumberKill, and NumberWound columns. It is difficult to predict these data. We can just fill in the average number of deaths in each country. But this will damage the data. We can also do the drop, but this will cause a lot of data to be lost even though we could have done an analysis not of NatTarget, NumberKill or NumberWound. Therefore, we should leave df2 like this. The data with data without missing value will be stored in another variable

In [ ]:
df_no_na = df2.dropna()

## **Change a Categorical Column to a Type**
Columns that have few unique values are usually categorical columns, we will change data types that have fewer than 25 columns to columns with data of type category

In [ ]:
cat_column = []
for i in range(len(df2.columns)):
    if df2.nunique()[i] < 25:
        cat_column.append(df2.nunique().index[i])
        df2[df2.nunique().index[i]] = df2[df2.nunique().index[i]].astype('category')
cat_column

['imonth',
 'extended',
 'crit1',
 'crit2',
 'crit3',
 'multiple',
 'region',
 'region_txt',
 'specificity',
 'attacktype1',
 'attacktype1_txt',
 'success',
 'suicide',
 'weaptype1',
 'weaptype1_txt',
 'targtype1',
 'targtype1_txt',
 'individual',
 'claimed',
 'property',
 'doubtterr']

In [ ]:
df2.to_csv('Data.csv')